### PR TITLE
[onert] Introduce LossMSEGrad function in train backend

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit introduces LossMSEGrad function in train backend.
    - Implements LossMSEGrad in cker
    - Removed usused derivative tensors

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Draft: #11035
Refer formula: https://github.com/Samsung/ONE/issues/11248#issuecomment-1674169085

### CODE REVIEW

File: compute/cker/include/cker/train/operation/Loss.h [hunk 1]
```diff
@@ -54,6 +54,23 @@ inline void MSE(const Shape &y_pred_shape, const T *y_pred_data, const Shape &y_
...
+    throw std::runtime_error("cker::MSEGrad: y_pred_shape != y_true_shape");
+  if (y_pred_shape != grad_shape)
+    throw std::runtime_error("cker::MSEGrad: y_pred_shape != grad_shape");
+
+  const auto y_pred = MapAsMatrixWithLastDimAsRows(y_pred_data, y_pred_shape);
+  const auto y_true = MapAsMatrixWithLastDimAsRows(y_true_data, y_true_shape);
+  auto grad = MapAsMatrixWithLastDimAsRows(grad_data, grad_shape);
+
+  auto size = y_pred.cols() * y_pred.rows();
+  grad = ((T)(-2) * (y_true - y_pred) / size);
```
Comment (ragmani) [inline compute/cker/include/cker/train/operation/Loss.h:71]:
```suggestion
  grad = (static_cast<T>(-2) * (y_true - y_pred) / size);
```

File: compute/cker/include/cker/train/operation/Loss.h [hunk 2]
```diff
@@ -54,6 +54,23 @@ inline void MSE(const Shape &y_pred_shape, const T *y_pred_data, const Shape &y_
...
+  if (y_pred_shape != y_true_shape)
+    throw std::runtime_error("cker::MSEGrad: y_pred_shape != y_true_shape");
+  if (y_pred_shape != grad_shape)
+    throw std::runtime_error("cker::MSEGrad: y_pred_shape != grad_shape");
+
+  const auto y_pred = MapAsMatrixWithLastDimAsRows(y_pred_data, y_pred_shape);
+  const auto y_true = MapAsMatrixWithLastDimAsRows(y_true_data, y_true_shape);
+  auto grad = MapAsMatrixWithLastDimAsRows(grad_data, grad_shape);
+
+  auto size = y_pred.cols() * y_pred.rows();
```
Comment (ragmani) [inline compute/cker/include/cker/train/operation/Loss.h:70]: If you want the number of all elements, you can use the code below.
```suggestion
  auto size = y_pred_shape.FlatSize();
```

Anyway I have a question.
Isn't it possible that the first dimension of  `y_pred` is batch size? If `y_pred` has a batch, shouldn't we calculate average per batch?

Comment (jyoungyun) [reply compute/cker/include/cker/train/operation/Loss.h:70]: As far as I know, the loss value is the sum or the mean over the batch. However, if user sets a Reduction option(?), user can get the loss of each batch size.
First of all, I didn't consider the Reduction case. How about considering it later? I will make another Grad function for this.

**Comment by jyoungyun** [comment]
@nnfw-bot test onert-x64-release
@nnfw-bot test onert-cross-release 👀 
@nnfw-bot test onert-cross-debug 👀

**Comment by jyoungyun** [comment]
@nnfw-bot test onert-x64-release

File: compute/cker/include/cker/train/operation/Loss.h (hunk 2 above)
Comment (jyoungyun) [reply compute/cker/include/cker/train/operation/Loss.h:70]: > Type of [tf.keras.losses.Reduction](https://www.tensorflow.org/api_docs/python/tf/keras/losses/Reduction) to apply to loss. Default value is AUTO. AUTO indicates that the reduction option will be determined by the usage context. For almost all cases this defaults to SUM_OVER_BATCH_SIZE.
> 
> from https://www.tensorflow.org/api_docs/python/tf/keras/losses/MeanSquaredError

Yes, you're right. We have to consider how to support other `Reduction` (I don't know whether we use the same name as tensorflow) options later. :)

Comment (ragmani) [reply compute/cker/include/cker/train/operation/Loss.h]: I didn't mean that we use the same name as tensorflow. :)

**Comment by hseok-oh** [comment]
```
[2023-08-14T01:37:23.139Z] /opt/jenkins_agent/workspace/nnfw/master/pr-onert-cross-debug/compute/cker/src/train/Loss.test.cc:146: Failure
[2023-08-14T01:37:23.139Z] Expected equality of these values:
[2023-08-14T01:37:23.139Z]   deriv_y_pred[i]
[2023-08-14T01:37:23.139Z]     Which is: 0.19999957
[2023-08-14T01:37:23.139Z]   expected[i]
[2023-08-14T01:37:23.139Z]     Which is: 0.2
[2023-08-14T01:37:23.139Z] [  FAILED  ] CKer_Operation.LossMSEGrad (0 ms)
[2023-08-14T01:37:23.139Z] [ RUN      ] CKer_Operation.neg_LossMSEGrad
[2023-08-14T01:37:23.139Z] [       OK ] CKer_Operation.neg_LossMSEGrad (0 ms)
[2023-08-14T01:37:23.139Z] [----------] 10 tests from CKer_Operation (0 ms total)
[2023-08-14T01:37:23.139Z] 
[2023-08-14T01:37:23.139Z] [----------] Global test environment tear-down
[2023-08-14T01:37:23.139Z] [==========] 10 tests from 1 test suite ran. (1 ms total)
[2023-08-14T01:37:23.139Z] [  PASSED  ] 9 tests.
[2023-08-14T01:37:23.139Z] [  FAILED  ] 1 test, listed below:
[2023-08-14T01:37:23.139Z] [  FAILED  ] CKer_Operation.LossMSEGrad
```

File: compute/cker/src/train/Loss.test.cc [hunk 3]
```diff
@@ -115,3 +115,87 @@ TEST(CKer_Operation, neg_LossMSE)
...
+    std::vector<float> y_pred = {1., 2., 3., 4., 5., 6., 7., 8., 9., 10.};
+    std::vector<float> y_true = {0., 1., 2., 3., 4., 5., 6., 7., 8., 9.};
+    std::vector<float> deriv_y_pred(10);
+    std::vector<float> expected = {0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2};
+
+    nnfw::cker::train::MSEGrad(nnfw::cker::Shape{1, 10}, y_pred.data(), nnfw::cker::Shape{1, 10},
+                               y_true.data(), nnfw::cker::Shape{1, 10}, deriv_y_pred.data());
+
+    for (size_t i = 0; i < deriv_y_pred.size(); ++i)
+      ASSERT_FLOAT_EQ(deriv_y_pred[i], expected[i]);
```
Comment (glistening) [inline compute/cker/src/train/Loss.test.cc:146]: We may use [`ASSERT_NEAR`](http://google.github.io/googletest/reference/assertions.html#EXPECT_NEAR).

**Comment by jyoungyun** [comment]
I fixed this issue.

> [2023-08-14T01:37:23.139Z] Expected equality of these values:
> [2023-08-14T01:37:23.139Z]   deriv_y_pred[i]
> [2023-08-14T01:37:23.139Z]     Which is: 0.19999957
> [2023-08-14T01:37:23.139Z]   expected[i]
> [2023-08-14T01:37:23.139Z]     Which is: 0.2

When I calculate this operation using Map itself, there is a difference in float operation result. When I changed it to element-wise operation, the result is exactly what I expected.

File: compute/cker/src/train/Loss.test.cc (hunk 3 above)
Comment (jyoungyun) [reply compute/cker/src/train/Loss.test.cc:146]: I fixed the code to get exactly the same result.

**Review by ragmani — APPROVED** [review]
LGTM

**Review by glistening — APPROVED** [review]
LGTM
